File: src/chapters/numerical-calculus/content/code/differentiation.py
import math


def deriv1(f, x, h=0.01):
    """Central-difference first derivative, error O(h^2)."""
    return (f(x + h) - f(x - h)) / (2 * h)


def deriv2(f, x, h=0.01):
    """Central-difference second derivative, error O(h^2)."""
    return (f(x + h) - 2 * f(x) + f(x - h)) / h ** 2


if __name__ == "__main__":
    print("f'(1)  ~", deriv1(math.sin, 1, 0.01), " exact cos(1)  =", math.cos(1))
    print("f''(1) ~", deriv2(math.sin, 1, 0.01), " exact -sin(1) =", -math.sin(1))

File: src/chapters/numerical-calculus/content/code/differentiation.js
// Central-difference first derivative, error O(h^2).
function deriv1(f, x, h = 0.01) {
  return (f(x + h) - f(x - h)) / (2 * h);
}

// Central-difference second derivative, error O(h^2).
function deriv2(f, x, h = 0.01) {
  return (f(x + h) - 2 * f(x) + f(x - h)) / (h * h);
}

console.log("f'(1)  ~", deriv1(Math.sin, 1, 0.01), " exact cos(1)  =", Math.cos(1));
console.log("f''(1) ~", deriv2(Math.sin, 1, 0.01), " exact -sin(1) =", -Math.sin(1));
